Fix empty text in components

RuscalWorld opened the pull request:
Quite a simple fix for some cases when we have JSON component that consists only of `extra` components, but does not have text itself. In such cases Notchian client expects `text` field to have `""` value, but when `Message` is serialized, this field is being omitted, which causes client to throw an exception.

Tnze commented:
```
Error:     message_test.go:117: encode Message error: get {"text":"","translate":"multiplayer.disconnect.server_full"} , want {"translate":"multiplayer.disconnect.server_full"}
```

Do notchian client expects text field to have "" value with translating message?

RuscalWorld commented:
Oh, I've completely forgotten about translations. It looks like `text` field takes priority over `translate`, so `{"text":"","translate":"multiplayer.disconnect.server_full"}` will not be rendered.

Any ideas on other ways of fixing this?..

RuscalWorld commented:
I think that `Message` can be splitted into two independent structures: one for translation components, and another one for text components. And both will have `translate` and `text` fields without `omitempty`.

RuscalWorld commented:
Now `Message` struct includes `MessageText` and `MessageTranslation` which contain `Text` and `Translate` fields. This breaks existing code where `Message` is initialized directly, but I've updated `Text` function that creates a `Message`, so it works like before. I've also added explicit detection of component type while unmarshalling JSON.

So, now it is possible to make components which contain empty `text` field and they are not conflicting with translations.

This code now passes all existing tests, I've just made minor changes to them to make them work with new API. I've also added some tests for marshalling and unmarshalling.

Tnze commented:
I have an idea.
We can implement `json.Marshaler` interface for `Message`. And override `MarshalJSON()`method, so we can tell wither if the `Message` is a translating message, and decide include or exclude the `text` property.

RuscalWorld commented:
> I have an idea. We can implement `json.Marshaler` interface for `Message`. And override `MarshalJSON()`method, so we can tell wither if the `Message` is a translating message, and decide include or exclude the `text` property.

I've already made it work like this 🤷‍♂️
https://github.com/Tnze/go-mc/blob/2e2ffdaa12028a0b2224b132bd2f601524446f02/chat/message.go#L102

Tnze commented:
No, we can fix this without breaking any other code.
